CRM-19657 - try to use systemwide default from address for recurring

jmcclelland opened the pull request:
receipts.

---

 * [CRM-19657: recurring contribution receipt comes from logged in user](https://issues.civicrm.org/jira/browse/CRM-19657)

seamuslee001 commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -4721,15 +4721,31 @@ public static function sendMail(&$input, &$ids, $contributionID, &$values, $recu
+      $userName = NULL;
+
+      // First, try to retrieve default name and email address.
+      $sitewide_default_from = array_pop(CRM_Core_OptionGroup::values('from_email_address', NULL, NULL, NULL, ' AND is_default = 1'));
```
@jmcclelland Jamie why not use the domain from email here rather than the default from the list of from emails?

jmcclelland replied:
OMG. That is way simpler and better. Thanks for the suggestion.

seamuslee001 commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -4721,15 +4721,28 @@ public static function sendMail(&$input, &$ids, $contributionID, &$values, $recu
+
+      if (empty($userEmail)) {
+        // If that didn't work, try to use the logged in user.
+        $session = CRM_Core_Session::singleton();
```
@jmcclelland Jamie this should probably be changed to $userID  = CRM_Core_Session::getLoggedInContactID()

jmcclelland replied:
Yes - also an excellent suggestion.

eileenmcnaughton commented:
I feel like I can agree with the logic for this - but I'm over merging anything that touches this function without a unit test. There is heaps of precedent in the api_v3_ContributionTest api though.

Note that where a contribution page id is attached I believe we need to prioritise the from etc from that page.

eileenmcnaughton commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -4721,15 +4721,27 @@ public static function sendMail(&$input, &$ids, $contributionID, &$values, $recu
-        $values['receipt_from_email'] = CRM_Utils_Array::value('receipt_from_email', $input, $userEmail);
-        $values['receipt_from_name'] = CRM_Utils_Array::value('receipt_from_name', $input, $userName);
+      // Set receipt from e-mail and name in value if we are sending the message.
+      $userEmail = NULL;
```
You don't really need to set these here. They prevent an enotice in theory but in practice there should be no e-notice from the code below so if they DID trigger an e-notice then it would be desirable because something unexpected would have happened

pradpnayak commented:
QA'd and commented under https://issues.civicrm.org/jira/browse/CRM-19657?focusedCommentId=97103&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-97103

eileenmcnaughton commented:
Thanks @pradpnayak looks like a really thorough round of test & review!

jmcclelland commented:
Yes - big thanks @pradpnayak

eileenmcnaughton commented:
@pradpnayak looks like Jamie has responded to your review - let me know if you think this is now mergeable

pradpnayak commented:
QA'd patched. Working fine
The from email address are take in order
1. civicrm_contribution_page.receipt_from_email
2. From Address mentioned under /civicrm/admin/options/from_email_address?reset=1
3. Logged in user's from email address

pradpnayak commented:
But still there is a bug in case of 
civicrm_contribution_page.receipt_from_email IS NULL, Domain Email Address IS NULL, Anonymous user

The from address is set to NULL. Should we use from_address of org email address from civicrm/admin/domain?action=update&reset=1 for such case?

eileenmcnaughton commented:
Thanks for the review - really clear notes! I agree with your analysis about a final fall back @jmcclelland ?

jmcclelland commented:
Sounds good to me. I just added that and re-factored a bit to make the logic more clear and rebased against the current 4.7.14 RC release to fix a conflict.

jmcclelland commented:
Oops. what have i done?

jmcclelland commented:
I really borked this one - opened new pull request with all commits flattened:

https://github.com/civicrm/civicrm-core/pull/9494

Sorry!
